File: bKap3_5_InSampleKitchenSink.m
% Script for performing the in-sample kitchen sink regression

% Clear console
clear; clc; close all;

% Set paths
sOldPath = path;
addpath('./Utils/');
sDataPath = './DATA/Liedtke/US/';
sResultsPath = './RESULTS/GWZ/';

%% Load data
% Macro panel produced by DATA/Liedtke/aggregate.py.
% CSV layout: col 1 = observation_date, col 2 = target (inflation),
%             col 3..end = predictors (already lag-aligned in Python).
tData     = readtable([sDataPath, 'aggregated.csv']);
cAllNames = tData.Properties.VariableNames;

% Parse dates (first column)
dtDates = tData{:,1};
if ~isdatetime(dtDates)
    dtDates = datetime(dtDates);
end

% Target (col 2) and predictors (col 3..end)
mX       = tData{:,3:end};          % predictors
vY       = tData{:,2};              % inflation target
cXnamesM = cAllNames(3:end)';       % predictor names as an Nx1 cell

%% Data preprocessing
% Predictive lag (periods). The predictors are already reporting-lag aligned
% in Python; this is the additional one-period predictive lag.
iTimeLag = 1;

% Determine dimensions
[iNumObs, iNumAssets]       = size(vY);
[iNumObsP, iNumPredictors]  = size(mX);

% Lag the predictors so that they are known one period ahead of the target
mXlag       = [NaN(iTimeLag, iNumPredictors); mX(1:end-iTimeLag,:)];

%% In-Sample analysis
% Get data / copy data
mXtemp      = mXlag;
vYtemp      = vY;
dtDatesTemp = dtDates;

% Remove missing values
lIsNaN = isnan(vYtemp) | any(isnan(mXtemp),2);
vYtemp(lIsNaN)      = [];
mXtemp(lIsNaN,:)    = [];
dtDatesTemp(lIsNaN) = [];

% z-Transformation of predictor
mXtemp = (mXtemp - mean(mXtemp,1))./std(mXtemp,[],1);

% Regression with intercept
rResults    = regstats(vYtemp, mXtemp, 'linear', {'tstat','rsquare'});
vBeta       = rResults.tstat.beta;      % coefficient estimates
vBetaT      = rResults.tstat.t;         % t-Statistic estimates
dR2         = rResults.rsquare;         % R2

% Add constant
mXtemp = [ones(size(mXtemp,1),1), mXtemp];

% In-sample prediction
mYhat = mXtemp * vBeta;

% Add missing values
mYhat = [NaN(sum(lIsNaN),1); mYhat];

% === Create table
% Merge all results and round to two digits
mResults = round([vBeta(2:end)*100, vBetaT(2:end)],2);

% Make to cell
cTable = [cXnamesM, sprintfc('%.2f', mResults)];

% Add column header
cTable = [{'Predictor','Beta', 'Beta T'}; cTable];

% Add R2
cTable = [cTable; cell(1, size(cTable,2)); {'R2','',num2str(round(dR2 * 100,2))}];

% === Save results
sFilename = [sResultsPath,'InSampleKitchenSinkResults.mat'];
save(sFilename, "cTable", 'mYhat');

% Restore path
path(sOldPath);

File: python/bKap3_5_InSampleKitchenSink.py
"""
bKap3_5_InSampleKitchenSink.py
Python port of bKap3_5_InSampleKitchenSink.m.

In-sample "kitchen-sink" regression: inflation on ALL (z-standardised)
predictors at once, using the complete-case sample. Reports beta / t-stat per
predictor and the overall R2, plus the in-sample fitted values.

Run from the repo root:  python python/bKap3_5_InSampleKitchenSink.py
"""

import os
import numpy as np
import pandas as pd
import scipy.io as sio

import models

DATA_PATH    = "./DATA/Liedtke/US/aggregated.csv"
RESULTS_PATH = "./RESULTS/GWZ/"

# =========================================================================
#  Load data
# =========================================================================
tData = pd.read_csv(DATA_PATH)
cAllNames = list(tData.columns)
dtDates = pd.to_datetime(tData.iloc[:, 0])
mX = tData.iloc[:, 2:].to_numpy(float)
vY = tData.iloc[:, 1].to_numpy(float)
cXnamesM = cAllNames[2:]

iTimeLag = 1
iNumObs, iNumPredictors = mX.shape

# Lag the predictors so they are known one period ahead of the target
mXlag = np.full_like(mX, np.nan)
mXlag[iTimeLag:, :] = mX[:-iTimeLag, :]

# =========================================================================
#  In-sample analysis (complete-case)
# =========================================================================
lIsNaN = np.isnan(vY) | np.isnan(mXlag).any(axis=1)
keep = ~lIsNaN
vYk = vY[keep]
mXk = mXlag[keep, :]

# z-transformation of each predictor column (sample std, ddof=1)
mXz = (mXk - mXk.mean(axis=0)) / mXk.std(axis=0, ddof=1)

reg = models.ols_regress(vYk, mXz)
vBeta = reg["beta"]      # [intercept, b1..bK]
vBetaT = reg["t"]
dR2 = reg["R2"]

# In-sample fitted values, NaN-padded at the top (removed rows are leading NaNs)
mYhat = np.concatenate([np.full(int(lIsNaN.sum()), np.nan), reg["yhat"]])

# =========================================================================
#  Build the result table (cTable) and save
# =========================================================================
mResults = np.round(np.column_stack([vBeta[1:] * 100, vBetaT[1:]]), 2)
tbl = pd.DataFrame(mResults, columns=["Beta", "Beta T"])
tbl.insert(0, "Predictor", cXnamesM)
print("========== In-sample kitchen-sink results ==========")
print(tbl.to_string(index=False))
print(f"\nR2 : {round(dR2 * 100, 2)}")

os.makedirs(RESULTS_PATH, exist_ok=True)
sio.savemat(os.path.join(RESULTS_PATH, "InSampleKitchenSinkResults_py.mat"), {
    "mYhat": mYhat.reshape(-1, 1), "vBeta": vBeta.reshape(-1, 1),
    "vBetaT": vBetaT.reshape(-1, 1), "dR2": dR2,
})
